File: execution/src/flow/task/calibration.ts
import { mkdir, rename, unlink, writeFile } from 'node:fs/promises';
import path from 'node:path';
import {
  applyStepPatch,
  pathExists,
  readFlowWithFingerprint,
  readJsonFile,
  resolveProjectFlow,
  taskProjectPaths,
  validateOverrides,
  validateProjectConfig,
  validateRoute,
} from './resolver.js';
import {
  CALIBRATION_PROPOSAL_SCHEMA_VERSION,
  type CalibrationHistoryRecord,
  type CalibrationProposal,
  type FlowOverrides,
  type FlowStepPatch,
  type TaskProjectConfig,
} from '../contracts/task-types.js';
import type { MidsceneFlowStep } from '../contracts/types.js';

export interface CalibrationOptions {
  project: string;
  proposal: string;
  projectRoot?: string;
}

export interface ValidatedCalibration {
  proposal: CalibrationProposal;
  proposalPath: string;
  currentSteps: Map<string, MidsceneFlowStep>;
  patchedSteps: Map<string, MidsceneFlowStep>;
  config: TaskProjectConfig;
  overrides: FlowOverrides;
}

function proposalFileName(proposal: string): string {
  const id = proposal.endsWith('.json') ? proposal.slice(0, -5) : proposal;
  if (!/^[a-z0-9][a-z0-9._-]*$/.test(id)) {
    throw new Error(`proposal id 只能包含小写字母、数字、点、下划线和连字符：${proposal}`);
  }
  return `${id}.json`;
}

function stableJson(value: unknown): string {
  return JSON.stringify(value);
}

function assertProposalShape(proposal: CalibrationProposal, expectedProject: string): void {
  if (proposal.schemaVersion !== CALIBRATION_PROPOSAL_SCHEMA_VERSION) {
    throw new Error(`不支持的 calibration proposal schemaVersion：${String(proposal.schemaVersion)}`);
  }
  if (proposal.project !== expectedProject) {
    throw new Error(`proposal 项目 ${proposal.project} 与请求项目 ${expectedProject} 不一致`);
  }
  if (!/^[a-z0-9][a-z0-9._-]*$/.test(proposal.id)) throw new Error(`无效 proposal id：${proposal.id}`);
  if (typeof proposal.summary !== 'string' || proposal.summary.trim() === '') throw new Error('proposal.summary 不能为空');
  if (typeof proposal.reason !== 'string' || proposal.reason.trim() === '') throw new Error('proposal.reason 不能为空');
  if (!/^[a-f0-9]{64}$/.test(proposal.baseFlowFingerprint)) throw new Error('proposal.baseFlowFingerprint 必须是 SHA-256');
  if (!Array.isArray(proposal.changes) || proposal.changes.length === 0) throw new Error('proposal.changes 不能为空');
}

export async function validateCalibrationProposal(options: CalibrationOptions): Promise<ValidatedCalibration> {
  const paths = taskProjectPaths(options.project, options.projectRoot);
  const proposalPath = path.join(paths.proposalsDir, proposalFileName(options.proposal));
  const [proposal, resolved, config, overrides, base] = await Promise.all([
    readJsonFile<CalibrationProposal>(proposalPath),
    resolveProjectFlow({ project: options.project, projectRoot: paths.projectRoot, executable: false }),
    readJsonFile<TaskProjectConfig>(paths.projectConfigPath),
    readJsonFile<FlowOverrides>(paths.overridesPath),
    readFlowWithFingerprint(paths.flowPath),
  ]);

  assertProposalShape(proposal, options.project);
  if (`${proposal.id}.json` !== path.basename(proposalPath)) {
    throw new Error(`proposal 文件名必须与 id 一致：${proposal.id}.json`);
  }
  if (proposal.baseFlowFingerprint !== base.fingerprint) {
    throw new Error('proposal 已过期：基础 IR 指纹已变化，请基于当前 flow 重新生成建议');
  }
  validateProjectConfig(config, base.flow);
  validateOverrides(overrides, base.flow);

  const currentSteps = new Map(resolved.flow.steps.map((step) => [step.id, step]));
  const patchedSteps = new Map<string, MidsceneFlowStep>();
  const changedIds = new Set<string>();
  for (const change of proposal.changes) {
    if (!change || typeof change !== 'object') throw new Error('proposal change 必须是对象');
    if (changedIds.has(change.stepId)) throw new Error(`proposal 重复修改 step：${change.stepId}`);
    changedIds.add(change.stepId);
    const current = currentSteps.get(change.stepId);
    if (!current) throw new Error(`proposal 引用了不存在的 step：${change.stepId}`);
    validateRoute(change.before?.route, `${change.stepId}.before.route`, false);
    if (stableJson(change.before.route) !== stableJson(current.route)) {
      throw new Error(`${change.stepId} 当前 route 与 proposal.before 不一致，请重新生成建议`);
    }
    if (change.before.timing !== undefined && stableJson(change.before.timing) !== stableJson(current.timing)) {
      throw new Error(`${change.stepId} 当前 timing 与 proposal.before 不一致，请重新生成建议`);
    }
    patchedSteps.set(change.stepId, applyStepPatch(current, change.after));
  }

  return { proposal, proposalPath, currentSteps, patchedSteps, config, overrides };
}

async function atomicWriteJson(filePath: string, value: unknown): Promise<void> {
  await mkdir(path.dirname(filePath), { recursive: true });
  const temporaryPath = `${filePath}.${process.pid}.tmp`;
  await writeFile(temporaryPath, `${JSON.stringify(value, null, 2)}\n`, 'utf8');
  await rename(temporaryPath, filePath);
}

function mergeAppliedPatch(existing: FlowStepPatch | undefined, patched: MidsceneFlowStep, after: FlowStepPatch): FlowStepPatch {
  const result: FlowStepPatch = { ...existing, route: structuredClone(patched.route) };
  if (after.timing) {
    result.timing = { waitBeforeMs: patched.timing?.waitBeforeMs ?? 0 };
  }
  return result;
}

export async function applyCalibrationProposal(options: CalibrationOptions): Promise<CalibrationHistoryRecord> {
  const validated = await validateCalibrationProposal(options);
  const paths = taskProjectPaths(options.project, options.projectRoot);
  const historyPath = path.join(paths.historyDir, `${validated.proposal.id}.json`);
  if (await pathExists(historyPath)) throw new Error(`校准历史已存在：${validated.proposal.id}`);

  const nextOverrides = structuredClone(validated.overrides);
  const nextConfig = structuredClone(validated.config);
  for (const change of validated.proposal.changes) {
    const patched = validated.patchedSteps.get(change.stepId)!;
    nextOverrides.steps[change.stepId] = mergeAppliedPatch(
      nextOverrides.steps[change.stepId],
      patched,
      change.after,
    );

    if (patched.route.strategy === 'input') {
      for (const definition of Object.values(nextConfig.inputs)) {
        if (definition.binding.stepId === change.stepId) definition.default = patched.route.value;
      }
    }
  }

  const history: CalibrationHistoryRecord = {
    ...validated.proposal,
    status: 'applied',
    appliedAt: new Date().toISOString(),
  };

  await atomicWriteJson(paths.overridesPath, nextOverrides);
  await atomicWriteJson(paths.projectConfigPath, nextConfig);
  await atomicWriteJson(historyPath, history);
  await unlink(validated.proposalPath);
  return history;
}

File: execution/cua/task/calibration.py
from __future__ import annotations

import re
from datetime import UTC, datetime
from pathlib import Path

from cua.domain.types import CalibrationOptions, ResolveProjectOptions, ValidatedCalibration
from cua.models.flow import InputRoute
from cua.models.task import (
    CalibrationHistoryRecord,
    CalibrationProposal,
    FlowOverrides,
    FlowStepPatch,
    TaskProjectConfig,
)
from cua.task.io import atomic_write_model, read_model
from cua.task.resolver import (
    apply_step_patch,
    read_flow_with_fingerprint,
    resolve_project_flow,
    task_project_paths,
    validate_overrides,
    validate_project_config,
)

PROPOSAL_ID_PATTERN = re.compile(r"^[a-z0-9][a-z0-9._-]*$")
FINGERPRINT_PATTERN = re.compile(r"^[a-f0-9]{64}$")


def proposal_file_name(proposal: str) -> str:
    proposal_id = proposal[:-5] if proposal.endswith(".json") else proposal
    if not PROPOSAL_ID_PATTERN.fullmatch(proposal_id):
        raise ValueError(f"proposal id 只能包含小写字母、数字、点、下划线和连字符：{proposal}")
    return f"{proposal_id}.json"


def validate_proposal_shape(proposal: CalibrationProposal, expected_project: str) -> None:
    if proposal.project != expected_project:
        raise ValueError(f"proposal 项目 {proposal.project} 与请求项目 {expected_project} 不一致")
    if not PROPOSAL_ID_PATTERN.fullmatch(proposal.id):
        raise ValueError(f"无效 proposal id：{proposal.id}")
    if not proposal.summary.strip():
        raise ValueError("proposal.summary 不能为空")
    if not proposal.reason.strip():
        raise ValueError("proposal.reason 不能为空")
    if not FINGERPRINT_PATTERN.fullmatch(proposal.base_flow_fingerprint):
        raise ValueError("proposal.baseFlowFingerprint 必须是 SHA-256")
    if not proposal.changes:
        raise ValueError("proposal.changes 不能为空")


def validate_calibration_proposal(options: CalibrationOptions) -> ValidatedCalibration:
    paths = task_project_paths(options.project, options.project_root)
    proposal_path = paths.proposals_dir / proposal_file_name(options.proposal)
    proposal = read_model(proposal_path, CalibrationProposal, "校准建议")
    resolved = resolve_project_flow(
        ResolveProjectOptions(project=options.project, project_root=paths.project_root, executable=False)
    )
    config = read_model(paths.project_config_path, TaskProjectConfig, "项目配置")
    overrides = read_model(paths.overrides_path, FlowOverrides, "校准配置")
    base_flow, fingerprint = read_flow_with_fingerprint(paths.flow_path)

    validate_proposal_shape(proposal, options.project)
    if f"{proposal.id}.json" != proposal_path.name:
        raise ValueError(f"proposal 文件名必须与 id 一致：{proposal.id}.json")
    if proposal.base_flow_fingerprint != fingerprint:
        raise ValueError("proposal 已过期：基础 IR 指纹已变化，请基于当前 flow 重新生成建议")
    validate_project_config(config, base_flow)
    validate_overrides(overrides, base_flow)

    current_steps = {step.id: step for step in resolved.flow.steps}
    patched_steps = {}
    changed_ids: set[str] = set()
    for change in proposal.changes:
        if change.step_id in changed_ids:
            raise ValueError(f"proposal 重复修改 step：{change.step_id}")
        changed_ids.add(change.step_id)
        current = current_steps.get(change.step_id)
        if current is None:
            raise ValueError(f"proposal 引用了不存在的 step：{change.step_id}")
        if change.before.route.to_json_dict() != current.route.to_json_dict():
            raise ValueError(f"{change.step_id} 当前 route 与 proposal.before 不一致，请重新生成建议")
        if change.before.timing is not None:
            current_timing = current.timing.to_json_dict() if current.timing else None
            if change.before.timing.to_json_dict() != current_timing:
                raise ValueError(f"{change.step_id} 当前 timing 与 proposal.before 不一致，请重新生成建议")
        patched_steps[change.step_id] = apply_step_patch(current, change.after)

    return ValidatedCalibration(
        proposal=proposal,
        proposal_path=proposal_path,
        current_steps=current_steps,
        patched_steps=patched_steps,
        config=config,
        overrides=overrides,
    )


def merge_applied_patch(
    existing: FlowStepPatch | None,
    patched_route: object,
    patched_wait_before_ms: int | None,
    include_timing: bool,
) -> FlowStepPatch:
    data = existing.to_json_dict() if existing else {}
    data["route"] = patched_route
    if include_timing:
        data["timing"] = {"waitBeforeMs": patched_wait_before_ms or 0}
    return FlowStepPatch.model_validate(data)


def apply_calibration_proposal(options: CalibrationOptions) -> CalibrationHistoryRecord:
    validated = validate_calibration_proposal(options)
    paths = task_project_paths(options.project, options.project_root)
    history_path = paths.history_dir / f"{validated.proposal.id}.json"
    if history_path.exists():
        raise ValueError(f"校准历史已存在：{validated.proposal.id}")

    next_overrides = validated.overrides.model_copy(deep=True)
    next_config = validated.config.model_copy(deep=True)
    for change in validated.proposal.changes:
        patched = validated.patched_steps[change.step_id]
        next_overrides.steps[change.step_id] = merge_applied_patch(
            next_overrides.steps.get(change.step_id),
            patched.route.to_json_dict(),
            patched.timing.wait_before_ms if patched.timing else None,
            change.after.timing is not None,
        )
        if isinstance(patched.route, InputRoute):
            for definition in next_config.inputs.values():
                if definition.binding.step_id == change.step_id:
                    definition.default = patched.route.value

    history = CalibrationHistoryRecord.model_validate(
        {
            **validated.proposal.to_json_dict(),
            "status": "applied",
            "appliedAt": datetime.now(UTC),
        }
    )
    atomic_write_model(paths.overrides_path, next_overrides)
    atomic_write_model(paths.project_config_path, next_config)
    atomic_write_model(history_path, history)
    validated.proposal_path.unlink()
    return history
